src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -51,28 +55,39 @@ public LocalLockManager getLocalLockManager()
         return null;
     }
 
-    FileBlock getNextBlock( final FileBlock prevBlock ) throws InterruptedException
+    FileBlock getNextBlock( final FileBlock prevBlock, final FileMeta metadata) throws IOException
     {
-        UUID next = prevBlock.getNextBlockID();
-        FileBlock nextBlock = blockCache.get( next );
-        if ( nextBlock == null )
+        synchronized ( InfinispanJFS.this )
         {
-            // setup a cache listener for this UUID, and wait in a timed loop for it to return
-            ClusterListener clusterListener = new ClusterListener( next );
-            clusterListener.listenToCache( blockCache );
-            // TODO: Figure out how to react to cache event
-            try
-            {
-                this.wait( 100 );
-            }
-            catch ( final InterruptedException e )
+            // TODO: make sure next will exist
+            UUID next = prevBlock.getNextBlockID();
+            FileBlock nextBlock = null;
+            while ( nextBlock == null )
             {
-                return null;
+                nextBlock = blockCache.get( next );
+                if (nextBlock == null)
+                {
+                    try
+                    {
+                        lockManager.reentrantSynchronous( metadata.getFilePath(), ( opLock ) -> {
+                            // setup a cache listener for this UUID, and wait in a timed loop for it to return
+                            ClusterListener clusterListener = new ClusterListener( next, opLock );
+                            clusterListener.listenToCache( blockCache );
+                            opLock.await( WAIT_TIMEOUT );
```

Comment:
I think you should be able to put this await operation inside the listenToCache() method. I think you should also be able to create the ClusterListener once, outside the while loop.

Reply:
that would mean putting the whole loop in a reentrantSynchronous lambda right?

Reply:
It looks like the right move is to push the whole thing here from the UUID to the return of nextBlock inside of the reentrantSynchronous lambda. This would do the path-specific locking that I was looking for, and eliminate the need for a synchronized block altogether. So...yes, you're right. :)

You would just need to return the return value from reentrantSynchronous in this approach.

Reply:
We also need a null check for next (UUID) since this is null on EOF blocks.